### ISSUE TITLE
Complete bulkification of lookup items for full recalc operations

### ISSUE THREAD

**Issue by jamessimone** [jamessimone/apex-rollup#107]
There are three rollup operations that are always full recalculations:

* first
* last
* average

Additionally, there are quite a few scenarios where a full recalc becomes part of the flow for the other operations. In these instances, there’s still some work to be done to properly bulkify Rollup - at present, it properly checks before starting each rollup operation if limits will be hit (and defers if so), but (and this is crucial for instances where more than 200 records would be supplied at any given time), _not_ after every lookup item’s rollup calculation.

There are some blue-sky changes that would be nice - notably, fully bulkifying the rollup calculations that involve full recalcs so that they only needed to be done once and applied to each lookup object (instead of one query per lookup object, which is what happens now, under the hood). If that ends up being too big to do at the moment, the short-term solution is to check after each parent item gets its rollup value supplied if limits are close to being hit (nice and easy; the deferral framework is already fully fleshed out), and splices the list of calc items accordingly so that only the unprocessed ones end up being deferred.
